### PR TITLE
Return `Ok` on kill if process has already exited

### LINKED ISSUE #112423: std::process::Child::kill should not fail if process is terminated on Windows
`Child::kill` calls `kill` on Unix, and `TerminateProcess` on Windows.

If process is terminated, but handle is alive yet, `kill` is successful on Unix, but fails on Windows.

`kill` is successful on Unix because the process is zombie until waited (it fails after process `waited`, that's probably OK)

According to [documentation of `TerminateProcess`](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess)

> After a process has terminated, call to TerminateProcess with open handles to the process fails with ERROR_ACCESS_DENIED (5) error code

So if someone else (not us) terminated the process, `kill` would fail.

There are problems with this behavior:
* API is different between Unix and Windows
* API to reliably terminate the process is complicated
  * call `kill`
  * if `kill` failed, need to `try_wait`. And if `try_wait` is successful, ignore the error of `kill`

I could reproduce it. I don't have easy access to Windows machine to create completely isolated test case, but repro looks like this:

```
let mut command = std::process::Command::new("powershell");
command.args(["-c", "Start-Sleep -Seconds 10000"]);
let mut child = command.spawn().unwrap();
let pid = child.id().try_into().unwrap();

// Terminate by process id, NOT by our process handle. Like if some other process killed it.
// This function calls `OpenProcess(pid)` followed by `TerminateProcess`.
kill(pid).unwrap();

thread::sleep(Duration::from_secs(5));

child.kill().expect("Failed to kill child process");
```

This fails with:

```
Failed to kill child process: Os { code: 5, kind: PermissionDenied, message: "Access is denied." }
```

exactly as described in WinAPI docs.

I think proper `kill` implementation should look like this:
* if process has been waited already, fail explicitly, to be consistent with Unix
* call `TerminateProcess`
* if `TerminateProcess` failed, call `GetExitCodeProcess`
* if `GetExitCodeProcess` is successful and exit code is not `STILL_ACTIVE`, consider `kill` successful

This is how `kill` is implemented in our project:

https://github.com/facebook/buck2/blob/b6fb1364e1caf6dac821767b27bda65726e38895/app/buck2_wrapper_common/src/winapi_process.rs#L58-L72

### Meta

`rustc --version --verbose`:
```
rustc 1.70.0-nightly
```
</p>
</details>

### PR DESCRIPTION
**Pull request by ChrisDenton**
This will require an FCP from @rust-lang/libs-api.

Fixes #112423. See that issue for more details.

### CODE REVIEW

**Comment by Amanieu** [comment]
@rfcbot fcp merge

**Comment by BurntSushi** [comment]
@rfcbot concern breaking-change

I am somewhat concerned by the breaking change here. The routine here explicitly documented that an `InvalidInput` error would be returned if `kill` was called on a process that had already terminated. I do agree that an `Ok(())` here probably makes more sense, and this seems like a sensible resolution to the linked bug report. But it is also a breaking change.

Do we have any user stories of folks that might be relying on the previous behavior that would be impacted by this change? My *sense* is that this is probably fine, but I think we should try to think about potential breakage here before making the change. A crater run might be helpful here too, although I'd be surprised if it provoked the necessary conditions to detect a breaking change here.

**Comment by ChrisDenton** [comment]
Yeah, the problem is that, despite the documentation, `InvalidInput` is only returned after a call to `wait` and even then only on Unix.

File: library/std/src/process.rs [hunk 1]
```diff
@@ -1904,8 +1904,8 @@ impl FromInner<imp::ExitCode> for ExitCode {
 }
 
 impl Child {
-    /// Forces the child process to exit. If the child has already exited, an [`InvalidInput`]
-    /// error is returned.
+    /// Forces the child process to exit. If the child has already exited, `Ok(())`
+    /// is returned.
```
Comment (stepancheg) [inline library/std/src/process.rs:1908]: A few lines below there's an example:

```
child.kill().expect("command wasn't running");
```

message is now incorrect: the only reason why `kill` fails is unknown system error.

File: library/std/src/sys/unix/process/process_unix.rs [hunk 2]
```diff
@@ -719,14 +719,15 @@ impl Process {
...
-            Err(io::const_io_error!(
-                ErrorKind::InvalidInput,
-                "invalid argument: can't kill an exited process",
-            ))
+            Ok(())
         } else {
-            cvt(unsafe { libc::kill(self.pid, libc::SIGKILL) }).map(drop)
+            match cvt(unsafe { libc::kill(self.pid, libc::SIGKILL) }).map(drop) {
+                // If the process can't be found or is already a zombie, return `Ok`.
+                Err(e) if e.raw_os_error() == Some(libc::ESRCH) => Ok(()),
```
Comment (stepancheg) [inline library/std/src/sys/unix/process/process_unix.rs:728]: This seems to be incorrect: it is fine to send `kill` to a zombie process, it is no-nop. And `ESRCH` actually indicates a serious problem: someone else has waited our process, so we should fail.

File: library/std/src/sys/windows/process.rs [hunk 3]
```diff
@@ -595,7 +595,14 @@ pub struct Process {
 
 impl Process {
     pub fn kill(&mut self) -> io::Result<()> {
-        cvt(unsafe { c::TerminateProcess(self.handle.as_raw_handle(), 1) })?;
+        let result = unsafe { c::TerminateProcess(self.handle.as_raw_handle(), 1) };
+        if result == c::FALSE {
+            let error = unsafe { c::GetLastError() };
+            // Return an error only if the process has not already died.
+            if error != c::ERROR_ACCESS_DENIED || self.try_wait().is_err() {
```
Comment (stepancheg) [inline library/std/src/sys/windows/process.rs:602]: Add a little more detailed comment please? Like this:

> `TerminateProcess` returns `ERROR_ACCESS_DENIED` if the process has been already terminated, so check if the process is terminated, and if so, do not return an error.

**Comment by BurntSushi** [comment]
@m-ou-se ah I get it now! Thanks for explaining a bit more. That makes sense.

@rfcbot resolve breaking-change

**Comment by tgross35** [comment]
If that was essentially a bug, could/should a regression test be added somehow?

**Comment by ChrisDenton** [comment]
Yep, a test sounds like a good idea. Give me a sec...

@bors r-

**Comment by ChrisDenton** [comment]
I've added a test and also changed the vxworks implementation to match the unix one.

@bors ready

**Comment by ChrisDenton** [comment]
Hm, does android not have echo? The `known_command` function appears to assume it does.

**Comment by Amanieu** [comment]
> Hm, does android not have echo? The `known_command` function appears to assume it does.

Android 13 definitely has `/system/bin/echo` (which is in `PATH`). However it may not be present on older versions, such as the Android emulator we are using in CI.

Perhaps you could try something using `shell_cmd` instead?

**Pull request merged**